include/aws/common/error.h

```diff
@@ -0,0 +1,104 @@
+#ifndef AWS_COMMON_ERROR_H_
+
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#define AWS_COMMON_ERROR_H_
+
+#include <aws/common/exports.h>
+#include <stdint.h>
+
+struct aws_error_info {
+    int error_code;
+    const char *literal_name;
+    const char *error_str;
+    const char *lib_name;
+    const char *formatted_name;
+};
+
+struct aws_error_info_list {
+    const struct aws_error_info *error_list;
+    uint16_t count;
+};
+
+#define AWS_DEFINE_ERROR_INFO(NAME, C, ES, LN)                                                                         \
+{                                                                                                                      \
+    .literal_name = #C,                                                                                                \
+    .error_code = C,                                                                                                   \
+    .error_str = ES,                                                                                                   \
+    .lib_name = LN,                                                                                                    \
+    .formatted_name = LN ": " #C ", " ES                                                                               \
+}
+
+typedef void(*aws_error_handler)(int err, void *ctx);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+ * Returns the latest error code on the current thread, or 0 if none have occurred.
+ */
+AWS_COMMON_API int aws_last_error(void);
+
+/*
+ * Returns the error str corresponding to `err`.
+ */
+AWS_COMMON_API const char *aws_error_str(int err);
+
+/*
+ * Returns the error lib name corresponding to `err`.
+ */
+AWS_COMMON_API const char *aws_error_lib_name(int err);
+
+/*
+ * Returns libname concatenated with error string.
+ */
+AWS_COMMON_API const char *aws_error_debug_str(int err);
+
+/*
+ * Raises `err` to the installed callbacks, and sets the thread's error.
+ */
+AWS_COMMON_API void aws_raise_error(int err);
```

Comment:
Have this return `-1` (or whatever the standard return code is for errored functions). This lets it be used in code like:

```
if (condition) return aws_raise_error(AWS_ERR_FOO);
```

Reply:
good idea.